app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -136,58 +159,105 @@ public void onBtnBackClick() {
     }
 
     public void onBtnNextClick() {
+        Slide slide = currentSlide();
         // Check if we are allowed to advance to the next slide.
-        if (mViewPager.getCurrentItem() == SLIDE_POS_LOCATION_PERMISSION) {
-            // As the storage permission is a prerequisite to run syncthing, refuse to continue without it.
-            if (!haveStoragePermission()) {
-                Toast.makeText(this, R.string.toast_write_storage_permission_required,
-                        Toast.LENGTH_LONG).show();
-                return;
-            }
+        switch (slide) {
+            case STORAGE:
+                // As the storage permission is a prerequisite to run syncthing, refuse to continue without it.
+                Boolean storagePermissionsGranted = PermissionUtil.haveStoragePermission(this);
+                if (!storagePermissionsGranted) {
+                    Toast.makeText(this, R.string.toast_write_storage_permission_required,
+                            Toast.LENGTH_LONG).show();
+                    return;
+                }
+                break;
+            case API_LEVEL_30:
+                if (!upgradedToApiLevel30()) {
+                    Toast.makeText(this, R.string.toast_api_level_30_must_reset,
+                            Toast.LENGTH_LONG).show();
+                    return;
+                }
+                break;
         }
 
-        int current = getItem(+1);
-        if (current < mLayouts.length) {
-            // Move to next slide.
-            mViewPager.setCurrentItem(current);
-            mBackButton.setVisibility(View.VISIBLE);
-        } else {
+        int next = mViewPager.getCurrentItem() + 1;
+        while (next < slideOrder.length) {
+            if (!shouldSkipSlide(slideOrder[next])) {
+                mViewPager.setCurrentItem(next);
+                mBackButton.setVisibility(View.VISIBLE);
+                break;
+            }
+            next++;
+        }
+        if (next == mLayouts.length) {
             // Start the app after "mNextButton" was hit on the last slide.
             Log.v(TAG, "User completed first start UI.");
             mPreferences.edit().putBoolean(Constants.PREF_FIRST_START, false).apply();
             startApp();
         }
     }
 
-    private void addBottomDots(int currentPage) {
-        mDots = new TextView[mLayouts.length];
+    private boolean isFirstStart() {
+        return mPreferences.getBoolean(Constants.PREF_FIRST_START, true);
+    }
 
-        int[] colorsActive = getResources().getIntArray(R.array.array_dot_active);
-        int[] colorsInactive = getResources().getIntArray(R.array.array_dot_inactive);
+    private boolean upgradedToApiLevel30() {
+        boolean upgraded = mPreferences.getBoolean(Constants.PREF_UPGRADED_TO_API_LEVEL_30,false);
```

Comment:
missing space after `,`, also could be folded into the if, given it's never reused.